File: Medium/Count Inversions/count-inversions.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution{
  public:
    // arr[]: Input Array
    // N : Size of the Array arr[]
    // Function to count inversions in the array.
    
    long long int merge(long long arr[], long long left, long long mid, long long right)
    {
        long long low = left;
        long long high = mid + 1;
        vector<long long int> temp;
        long long int inversion = 0;
        
        while(low <= mid && high <= right)
        {
            if(arr[low] <= arr[high])
            {
                temp.emplace_back(arr[low]);
                low++;
            }
            else
            {
                inversion += (mid + 1 - low);
                temp.emplace_back(arr[high]);
                high++;
            }
        }
        
        while(low <= mid)
        {
            temp.emplace_back(arr[low]);
            low++;
        }
        
        while(high <= right)
        {
            temp.emplace_back(arr[high]);
            high++;
        }
        
        for(int i = left; i <= right ; i++)
        {
            arr[i] = temp[i - left];
        }
        
        return inversion;
    }
    
    
    long long int mergesort(long long arr[],long long left, long long right)
    {
        long long int inversion = 0;
        if(left < right)
        {
            long long mid = (left + right)/2;
            inversion += mergesort(arr,left,mid);
            inversion += mergesort(arr,mid+1,right);
            inversion += merge(arr,left,mid,right);
        }
        
        return inversion;
    }
    
    
    
    long long int inversionCount(long long arr[], long long N)
    {
        // Your Code Here

        long long int inversions = mergesort(arr,0,N-1);
        return inversions;
    }

};

//{ Driver Code Starts.

int main() {
    
    long long T;
    cin >> T;
    
    while(T--){
        long long N;
        cin >> N;
        
        long long A[N];
        for(long long i = 0;i<N;i++){
            cin >> A[i];
        }
        Solution obj;
        cout << obj.inversionCount(A,N) << endl;
    }
    
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Count Inversions/count-inversions.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Sorting {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        for (int g = 0; g < t; g++) {
            String[] str = (br.readLine()).trim().split(" ");
            int arr[] = new int[str.length];
            for (int i = 0; i < str.length; i++) arr[i] = Integer.parseInt(str[i]);
            System.out.println(new Solution().inversionCount(arr));
            System.out.println("~");
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    // Function to count inversions in the array.
    
    public static int merge(int low, int mid, int high, int arr[])
    {
        int left = low;
        int right = mid + 1;
        List<Integer> list = new ArrayList<>();
        
        int cnt = 0;
        
        while(left <= mid && right <= high)
        {
            if(arr[left] <= arr[right])
            {
                list.add(arr[left]);
                left++;
            }
            else
            {
                cnt += (mid - left + 1);
                list.add(arr[right]);
                right++;
            }
        }
        
        while(left <= mid)
        {
            list.add(arr[left]);
            left++;
        }
        
        while(right <= high)
        {
            list.add(arr[right]);
            right++;
        }
        
        for(int i = low; i <= high; i++)
        {
            arr[i] = list.get(i - low);
        }
        
        return cnt;
    }
    
    
    public static int msort(int low, int high, int arr[])
    {
        int cnt = 0;
        
        if(low >= high) return cnt;
        
        int mid = (low + high) >> 1;
        cnt += msort(low, mid, arr);
        cnt+= msort(mid+1, high, arr);
        cnt += merge(low, mid, high, arr);
        return cnt;
    }
    
    
    static int inversionCount(int arr[]) {
        // Your Code Here
        int n = arr.length;
        
        return msort(0, n - 1, arr);
        
    }
}